tests/federation_room_join_partial_state_test.go

```diff
@@ -3295,6 +3295,61 @@ func TestPartialStateJoin(t *testing.T) {
 		}
 	})
 
+	t.Run("Room stats are correctly updated once state re-sync completes", func(t *testing.T) {
+		deployment := Deploy(t, b.BlueprintAlice)
+		defer deployment.Destroy(t)
+
+		// create a user with admin powers as we will need this power to make the remote room visible in the
+		// local room list
+		terry := deployment.RegisterUser(t, "hs1", "terry", "pass", true)
+
+		server := createTestServer(t, deployment)
+		cancel := server.Listen()
+		defer cancel()
+		serverRoom := createTestRoom(t, server, terry.GetDefaultRoomVersion(t))
+
+		// start a partial state join
+		psjResult := beginPartialStateJoin(t, server, serverRoom, terry)
+		defer psjResult.Destroy(t)
+
+		// make the remote room visible in the local room list
+		reqBody := client.WithJSONBody(t, map[string]interface{}{
+			"visibility": "public",
+		})
+		terry.MustDoFunc(t, "PUT", []string{"_matrix", "client", "v3", "directory", "list", "room", serverRoom.RoomID}, reqBody)
+
+		// sanity check - before the state has completed syncing state we would expect only one user
+		// to show up in the room list
+		res := terry.MustDoFunc(t, "GET", []string{"_matrix", "client", "v3", "publicRooms"})
+		body, err := ioutil.ReadAll(res.Body)
+		if err != nil {
+			t.Fatalf("something broke: %v", err)
+		}
+		numJoinedMembers := gjson.GetBytes(body, "chunk.0.num_joined_members")
+		if numJoinedMembers.Int() != 1 {
+			t.Fatalf("Expected 1 users, returned %s", numJoinedMembers)
+		}
```

Comment:
I think Kegan's intention for complement is to make more use of the `must` and `match` libraries. You might be able to use `must.MatchResponse` here to make it neater, as in e.g.

https://github.com/matrix-org/complement/blob/ec6e09c17eddb6c0b4686c76c3a2f7272cb708f2/tests/csapi/room_members_test.go#L175-L187

something like 

```go
must.matchResponse(t, res, match.HTTPResponse{
    StatusCode: 200,
    JSON: []match.JSON{
        match.JSONKeyEqual("chunk.0.num_joined_members", 1)
    }
})

I don't know if you find that clearer. It's a little more declarative and delegates the error handling I guess. :shrug: